packages/runtime-dom/src/components/Transition.ts

```diff
@@ -136,24 +137,40 @@ export function resolveTransitionProps(
     }
   }
 
+  // see issue #2531, #2593
+  let cacheTransition: string
+  const disableTransition = (el: Element) => {
+    const style = (el as HTMLElement).style
+    cacheTransition = style.transition
+    style.transitionProperty = 'none'
+  }
+
+  const enableTransition = (el: Element) => {
+    ;(el as HTMLElement).style.transition = cacheTransition
+  }
+
   return extend(baseProps, {
     onBeforeEnter(el) {
       onBeforeEnter && onBeforeEnter(el)
       addTransitionClass(el, enterActiveClass)
       addTransitionClass(el, enterFromClass)
+      disableTransition(el)
     },
     onBeforeAppear(el) {
       onBeforeAppear && onBeforeAppear(el)
       addTransitionClass(el, appearActiveClass)
       addTransitionClass(el, appearFromClass)
+      disableTransition(el)
     },
     onEnter: makeEnterHook(false),
     onAppear: makeEnterHook(true),
     onLeave(el, done) {
       const resolve = () => finishLeave(el, done)
       addTransitionClass(el, leaveActiveClass)
       addTransitionClass(el, leaveFromClass)
+      disableTransition(el)
```

Comment:
Given that this is only needed for the leave transition, the logic should simply be inlined here.